### PR TITLE
Add PyOp doc and UT

### PR DESCRIPTION
**Pull request by RandySheriffH**
Document python operator and enable the UT in ci pipelines.

### CODE REVIEW

**Comment by pranavsharma** [comment]
Please add readable title and description for the PR. May be something like - Add documentation for PyOp.

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 1]
```diff
@@ -218,7 +218,7 @@ TEST_F(CApiTest, custom_op_handler) {
   TestInference<PATH_TYPE>(env_, CUSTOM_OP_MODEL_URI, inputs, "Y", expected_dims_y, expected_values_y, 0, custom_op_domain);
 }
 
-#ifdef ENABLE_LANGUAGE_INTEROP_OPS
+#if defined(ENABLE_LANGUAGE_INTEROP_OPS) && !( defined(_WIN32) && defined(_DEBUG))
```
Comment (pranavsharma) [inline onnxruntime/test/shared_lib/test_inference.cc]: Can you add a helpful comment here stating why this logic is required?

File: tools/ci_build/build.py [hunk 2]
```diff
@@ -579,6 +579,39 @@ def run_onnx_tests(build_dir, configs, onnx_test_data_dir, provider, enable_para
           run_subprocess([exe,'-x'] + cmd, cwd=cwd)
 
 
+# mkldnn temporary function for running onnx tests and model tests separately.
+def mkldnn_run_onnx_tests(build_dir, configs, onnx_test_data_dir):
```
Comment (pranavsharma) [inline tools/ci_build/build.py]: Is this is because of rebasing with master? Doesn't look like related to this PR.

File: docs/PyOp.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Python Operator
```
Comment (pranavsharma) [inline docs/PyOp.md]: I would introduce this as Support for Language Interop and include the python operator as a sub-heading. Also, can you mark this as beta? like Python Operator (Beta).

File: docs/PyOp.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Python Operator 
+## Introduction
+To facilitate Python coders on model developing, onnxruntime provides a way to invoke operators implemented in Python.
+To use the feature, model designer needs to:
+1. Define Python operator node with all required attributes;
+2. Implement a Python module of all referred operators in required format;
+3. Place the Python module into python sys path, then do referencing with onnxruntime.
+
+## Usage
+Currently, Python operator is only published via source under onnxruntime/core/language_interop_ops, developers need to compile with“-enable_language_interop_ops”and override existing onnxruntime binary. Meanwhile, please also copy onnxruntime_pywrapper.dll or libonnxruntime_pywrapper.so or libonnxruntime_pywrapper.dylib to the path where onnxruntime binary is placed.
```
Comment (pranavsharma) [inline docs/PyOp.md]: Can you write this step by step? Like step 1- do this. step 2 - do that. etc.

File: cmake/onnxruntime_pyop.cmake [hunk 5]
```diff
@@ -1,30 +1,45 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-FIND_PACKAGE(PythonLibs)
```
Comment (pranavsharma) [inline cmake/onnxruntime_pyop.cmake]: can you add a one line comment before each of the exec_programs saying what it is intending to achieve?

File: tools/ci_build/build.py (hunk 2 above)
Comment (RandySheriffH) [reply tools/ci_build/build.py]: Exactly, I have cancelled the change.

File: onnxruntime/test/shared_lib/test_inference.cc (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/test/shared_lib/test_inference.cc]: Adding comments.

File: docs/PyOp.md (hunk 4 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Sure, adding steps.

File: docs/PyOp.md (hunk 3 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Indeed, logically we should orgnize them in the tree like:

InterOp
   |-------PyOP
   |-------CsOp
   |-------JavaOP

But would it be little bit overwhelming for clients who cares exclusively on either PyOp, CsOp or JavaOp?

File: cmake/onnxruntime_pyop.cmake (hunk 5 above)
Comment (RandySheriffH) [reply cmake/onnxruntime_pyop.cmake]: Add comments.

File: docs/PyOp.md (hunk 3 above)
Comment (pranavsharma) [reply docs/PyOp.md]: Ok, let's keep it as is.

**Comment by pranavsharma** [comment]
Looks ok. can you fix the builds?

File: docs/PyOp.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+Step 3, implement mymodule.py:
+```python
+class Multi_1:
+    def __init__(self, W1, W2, W3):
+        self.W1 = int(W1)
+        self.W2 = int(W2)
+        self.W3 = int(W3)
+    def compute(self, S, P):
+        ret = S + P
+        return ret + self.W1, ret + self.W2, ret + self.W3
```
Comment (xadupre) [inline docs/PyOp.md]: What is the syntax if there is only one result?

File: docs/PyOp.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+* TensorProto.BOOL,
+* TensorProto.UINT8,
+* TensorProto.UINT16,
+* TensorProto.UINT32,
+* TensorProto.INT16,
+* TensorProto.INT32,
+* TensorProto.FLOAT,
+* TensorProto.DOUBLE
+
+## Limitations
```
Comment (xadupre) [inline docs/PyOp.md]: What happens if the script requires an additional file (like a model to load?)

File: docs/PyOp.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
+# Python Operator 
+To facilitate Python coders on model developing, onnxruntime provides a way to invoke operators implemented in Python.
+
+## Usage
+Step 1, build onnxruntime with“--config Release --enable_language_interop_ops --build_shared_lib”and override existing onnxruntime binary with the latest, then copy onnxruntime_pywrapper.dll or libonnxruntime_pywrapper.so or libonnxruntime_pywrapper.dylib to the path where onnxruntime binary is placed.
```
Comment (xadupre) [inline docs/PyOp.md]: There is also a flag to avoid the build to install a different version of numpy and build with the current one. Maybe that's Worth mentioning in this case.

Comment (wenbingl) [inline docs/PyOp.md]: Can we reuse the enable_python_bindings to enable this capability? The current config in the command line looks pretty irrelevant.

File: docs/PyOp.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+Note that it is suggested to compile within the Python environment where inferencing will happen. For example, if inferencing will happen in a conda env named myconda1, please compile the binary within that environment as well.
```
Comment (wenbingl) [inline docs/PyOp.md]: A simple figure to show how the pyop is interactive with onnxruntime will much help the end user to get the while idea.

File: docs/PyOp.md (hunk 6 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: just return it:

return <your result>

File: docs/PyOp.md (hunk 7 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Just import it as we normally do within a python file.

File: docs/PyOp.md (hunk 8 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: OK, but which flag you mean exactly?

Comment (RandySheriffH) [reply docs/PyOp.md]: The purpose of adding new flags is to make sure user know exactly what he wants.
Python bindings does not necessarily mean a user is expecting pyop.
Conversely, pyop does not require python binding at all.

File: docs/PyOp.md (hunk 9 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Good suggestion, let me add it.

File: docs/PyOp.md (hunk 6 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Coders are allowed to return arbitrary number of numpys.

Comment (xadupre) [reply docs/PyOp.md]: I was asking because looking at the example, I was wondering if it should be a tuple with one element or the element itself knowing that this element cannot be a tuple then.

File: docs/PyOp.md (hunk 8 above)
Comment (xadupre) [reply docs/PyOp.md]: --numpy_version= with no value

Comment (xadupre) [reply docs/PyOp.md]: I think it is better this way, otherwise what would happen with the C or C# API?

File: docs/PyOp.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+## Limitations
+* On Windows,  "--config Debug" has known issues,  build with "--config RelWithDebInfo" if need debugging symbols;
+* Due to python C API restrictions, multi-threading is disabled, meaning Python operators will run sequentially.
+
+## Test
+The operator has been tested on multiple platforms, with or without conda:
+
+Platform | Python 3.5 | Python 3.6 | Python 3.7
+----------- | ------------| -----------  | -----------
+Windows | (conda) passed | (conda) passed | passed
```
Comment (xadupre) [inline docs/PyOp.md]: Quick question: did we try if it works if the python script import pytorch, Tensorflow or spacy?

Comment (RandySheriffH) [reply docs/PyOp.md]: No, did not test the part.

File: docs/PyOp.md (hunk 8 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Yes, it would happend in C# API - if someone is using C# API and he want to also go with PyOP.

File: docs/PyOp.md (hunk 6 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: If only one numpy returned, Python C API will simply return an Array.

File: docs/PyOp.md (hunk 8 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Added a note for that. Let's if it is good enough.

File: docs/PyOp.md (hunk 10 above)
Comment (RandySheriffH) [reply docs/PyOp.md]: Just test some cases with imported tensorflow, all fine.
